code/datums/storage/storage.dm

```diff
@@ -0,0 +1,417 @@
+
+/* Looking to add a storage to an item? */
+// Just call atom.create_storage() as below
+// Ex. myatom.create_storage(/datum/storage) creates a very simple storage on an item.
+// Procs for common storage actions, like getting all contents from a storage, are available.
+// The storage datum is organized into "internal" and "public" procs
+//
+// To remove a storage from an item, just call atom.remove_storage().
+
+/* When should a new storage datum type be created? */
+// A new storage datum type should be made if you are looking to override the code for adding/removing an item to a storage, or make a significant
+// change that needs a new type. Ex. See bible storages.
+
+/* Lower level things to note */
+// -Items in a storage are kept track of in the stored item's contents list and in the storage datum, but when iterating through contents, the procs
+// in this file should be used for getting the contents
+// -For /obj/item, src.loc and src.stored.linked_item will be the same, but the second should be used for consistency and clarity
+
+/* Examples */
+// -Storage datums on clothing items function just like backpacks
+
+// add storage to an item
+/atom/proc/create_storage(storage_type, list/spawn_contents = list(), list/can_hold = list(), list/can_hold_exact = list(), check_wclass = FALSE,
+		max_wclass = W_CLASS_SMALL, slots = 7, sneaky = FALSE, opens_if_worn = FALSE)
+	var/list/previous_storage = list()
+	for (var/obj/item/I as anything in src.storage?.get_contents())
+		previous_storage += I
+	src.remove_storage()
+	src.storage = new storage_type(src, spawn_contents, can_hold, can_hold_exact, check_wclass, max_wclass, slots, sneaky, opens_if_worn)
+	for (var/obj/item/I as anything in previous_storage)
+		src.storage.add_contents(I)
+
+// remove an atom's storage
+/atom/proc/remove_storage()
+	qdel(src.storage)
+	src.storage = null
+
+/datum/storage
+	/// Types that can be held
+	var/list/can_hold = null
+	/// Exact types that can be held, in addition to can_hold, if it has types
+	var/list/can_hold_exact = null
+	/// If set, if can_hold is used, an item not in can_hold or can_hold_exact can fit in the storage if its weight is low enough
+	var/check_wclass = FALSE
+	/// Storage hud attached to the storage
+	var/datum/hud/storage/hud = null
+	/// Don't print a visible message on use
+	var/sneaky = FALSE
+	/// Prevent accessing storage when clicked when worn, ex. in pocket
+	var/opens_if_worn = FALSE
+	/// Maximum w_class that can be held
+	var/max_wclass = W_CLASS_SMALL
+	/// Number of storage slots, even numbers overlap the close button for the on-ground hud layout
+	var/slots = 7
+	/// Does moving the linked storage item cause anything to happen to stored items
+	var/move_triggered = TRUE
+	/// The storage item linked to this datum
+	var/atom/linked_item = null
+	/// All items stored
+	var/list/stored_items = list()
```

Comment:
initialize this in New please

Reply:
or lazily

https://hackmd.io/@goonstation/code#StartupRuntime-trade-offs-with-lists-and-the-%E2%80%9Chidden%E2%80%9D-init-proc

Reply:
Put in New()